### PR TITLE
feat(gatsby,gatsby-cli): Pass an errorMap to reporter.error

### PR DESCRIPTION
**Pull request by abhiaiyer91**
This PR has enhancements to the reporter (in gatsby cli) and "local reporter" (in gatsby).

The Reporter Class has added
`errorMap` -> Plugins can register new structured errors in `onPreInit` by calling `reporter.setErrorMap`

### CODE REVIEW

File: packages/gatsby-cli/src/reporter/reporter.ts [hunk 1]
```diff
@@ -36,6 +41,42 @@ class Reporter {
...
+    const entries = Object.entries(entry)
+
+    const newErrorMap = entries.reduce((memo, [key, val]) => {
+      memo[`${name}_${key}`] = val
+
+      return memo
+    }, {})
+
+    this.setErrorMap(newErrorMap)
+  }
```
Comment (wardpeet) [inline packages/gatsby-cli/src/reporter/reporter.ts:70]: do we want to expose it on the reporter? Maybe it's better to export a utility that we use in api-runner. It feels like a private function

File: packages/gatsby-cli/src/reporter/reporter.ts [hunk 2]
```diff
@@ -136,7 +195,8 @@ class Reporter {
       }
     }
 
-    const structuredError = constructError({ details })
+    const structuredError = constructError({ details }, this.getErrorMap())
```
Comment (wardpeet) [inline packages/gatsby-cli/src/reporter/reporter.ts:198]: Can we do the errorMap dance here so we don't have to spread each time?

```suggestion
    let errorMapEntry = defaultError;
    if (details.id) {
      // Look at original errorMap, ids cannot be overwritten
      if (errorMap[details.id]) {
        errorMapEntry = errorMap[details.id]
      }
  
      if (this.errorMap[details.id]) {
        errorMapEntry = this.errorMap[details.id]
      }
    }
    
    const structuredError = constructError({ details }, errorMapEntry)
```

File: packages/gatsby-cli/src/reporter/reporter.ts [hunk 3]
```diff
@@ -72,6 +113,24 @@ class Reporter {
...
+    if (typeof errorMeta === `object`) {
+      const id = errorMeta && errorMeta[`id`]
+
+      if (id) {
+        errorMeta[`id`] = `${pluginName}_${id}`
+      }
+    }
+
+    this[methodName](errorMeta, error)
+  }
```
Comment (wardpeet) [inline packages/gatsby-cli/src/reporter/reporter.ts:132]: Same her do we want it on the reporter and not as a utility? I want to limit the public api from reporter.

**Review by wardpeet — COMMENTED** [review]
This looks great! I'm mostly wondering about if we want to make the reporter public API broader.

File: packages/gatsby-cli/src/reporter/reporter.ts (hunk 3 above)
Comment (abhiaiyer91) [reply packages/gatsby-cli/src/reporter/reporter.ts:132]: utility is good

File: packages/gatsby-cli/src/structured-errors/__tests__/construct-error.ts [hunk 4]
```diff
@@ -34,7 +36,23 @@ test(`it logs error on invalid schema`, () => {
...
+        text: (context): string => `Error text is ${context.someProp} `,
+        level: Level.ERROR,
+        docsUrl: `https://www.gatsbyjs.org/docs/gatsby-cli/#new`,
+      },
+    }
+  )
+
+  expect(error.code).toBe(`1337`)
+  expect(error.docsUrl).toBe(`https://www.gatsbyjs.org/docs/gatsby-cli/#new`)
+})
```
Comment (wardpeet) [inline packages/gatsby-cli/src/structured-errors/__tests__/construct-error.ts:58]: Can you add a test so we check if original errorMap cannot be overridden.

File: packages/gatsby-cli/src/structured-errors/construct-error.ts [hunk 5]
```diff
@@ -1,25 +1,37 @@
...
+  if (id) {
+    // Look at original errorMap, ids cannot be overwritten
+    if (errorMap[id]) {
+      errorMapEntry = errorMap[id]
+    }
+
+    if (suppliedErrorMap[id]) {
+      errorMapEntry = suppliedErrorMap[id]
+    }
+  }
```
Comment (wardpeet) [inline packages/gatsby-cli/src/structured-errors/construct-error.ts:23]: It's probably my mistake in a previous comment but it should be
```suggestion
  if (id) {
    // Look at original errorMap, ids cannot be overwritten
    if (errorMap[id]) {
      errorMapEntry = errorMap[id]
    } else if (suppliedErrorMap[id]) {
      errorMapEntry = suppliedErrorMap[id]
    }
  }
```

File: packages/gatsby/src/utils/api-runner-node.js [hunk 6]
```diff
@@ -115,10 +115,128 @@ const deferActions = actions => {
...
+
+function extendErrorIdWithPluginName(pluginName, errorMeta) {
+  if (typeof errorMeta === `object`) {
+    const id = errorMeta && errorMeta[`id`]
+
+    if (id) {
+      errorMeta[`id`] = `${pluginName}_${id}`
+    }
+  }
+}
```
Comment (wardpeet) [inline packages/gatsby/src/utils/api-runner-node.js:140]: Let's keep this side-effect free so we don't augment the object but instead create a new one.

```suggestion
function extendErrorIdWithPluginName(pluginName, errorMeta) {
  if (typeof errorMeta === `object`) {
    const id = errorMeta && errorMeta[`id`]

    if (id) {
      return {
        ...errorMeta,
        id: `${pluginName}_${id}`
      }
    }
  }
  
  return errorMeta;
}
```

File: packages/gatsby/src/utils/api-runner-node.js [hunk 7]
```diff
@@ -115,10 +115,128 @@ const deferActions = actions => {
...
+      extendErrorIdWithPluginName(pluginName, errorMeta)
+
+      return reporter.panic(errorMeta, error)
+    }
+
+    panicOnBuild = (errorMeta, error) => {
+      extendErrorIdWithPluginName(pluginName, errorMeta)
+
+      return reporter.panicOnBuild(errorMeta, error)
+    }
```
Comment (wardpeet) [inline packages/gatsby/src/utils/api-runner-node.js:186]: With the above change we should be able to move to

```suggestion
    error = (errorMeta, error) => {    
      return reporter.error(extendErrorIdWithPluginName(pluginName, errorMeta), error)
    }

    panic = (errorMeta, error) => {
      return reporter.panic(extendErrorIdWithPluginName(pluginName, errorMeta), error)
    }

    panicOnBuild = (errorMeta, error) => {
      return reporter.panicOnBuild(extendErrorIdWithPluginName(pluginName, errorMeta), error)
    }
```

File: packages/gatsby-cli/src/reporter/reporter.ts [hunk 8]
```diff
@@ -36,6 +37,27 @@ class Reporter {
...
+      ...entry,
+    }
+  }
+
+  /**
+   * Retrieve error map
+   */
+  getErrorMap = (): Record<ErrorId, IErrorMapEntry> =>
+    // TODO: We always spread the internal error map to ensure our keys do not get overwritten
+    this.errorMap
```
Comment (wardpeet) [inline packages/gatsby-cli/src/reporter/reporter.ts:59]: Let's not expose this function, so users can't augment it themselves.
```suggestion
```

File: packages/gatsby-cli/src/reporter/reporter.ts [hunk 9]
```diff
@@ -136,7 +158,8 @@ class Reporter {
       }
     }
 
-    const structuredError = constructError({ details })
+    const structuredError = constructError({ details }, this.getErrorMap())
```
Comment (wardpeet) [inline packages/gatsby-cli/src/reporter/reporter.ts:161]:
```suggestion
    const structuredError = constructError({ details }, this.errorMap)
```

**Review by wardpeet — COMMENTED** [review]
Added some nitpicks to make code side-effect free

File: packages/gatsby-cli/src/structured-errors/error-schema.ts [hunk 10]
```diff
@@ -20,6 +20,7 @@ export const errorSchema: Joi.ObjectSchema<IStructuredError> = Joi.object().keys
         })
       )
       .allow(null),
+    errorCategory: Joi.string().valid([`USER`, `SYSTEM`]),
```
Comment (wardpeet) [inline packages/gatsby-cli/src/structured-errors/error-schema.ts:23]: it's already part of an error, should we prefix it with error?
```suggestion
    category: Joi.string().valid([`USER`, `SYSTEM`]),
```

**Comment by LekoArts** [comment]
Some questions around this (don't know if it was discussed already):

1) We use 5-digit numbers for the errors so that those are better searchable in Google than e.g. `1` as an ID. Can/Should we enforce the same format for plugin errors?
2) Since plugins can't override already existing errors from our error map we should at least show a warning somewhere I guess? Otherwise the error reporting silently fails/displays the wrong error message.
3) Our error map has a loose format, e.g. all GraphQL errors are [prefixed with 859](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/structured-errors/error-map.ts#L49-L257) and we're currently at 85927.
	- What happens when a plugin developer decides to claim 85928 and we add the same ID sometime later to our own errorMap?
	- What happens when two plugins have the same ID?
	- Should we enforce certain ranges for our internal errorMap and the external?

File: packages/gatsby-cli/src/structured-errors/error-schema.ts [hunk 11]
```diff
@@ -20,6 +20,7 @@ export const errorSchema: Joi.ObjectSchema<IStructuredError> = Joi.object().keys
         })
       )
       .allow(null),
+    category: Joi.string().valid([`USER`, `SYSTEM`]),
```
Comment (julienp) [inline packages/gatsby-cli/src/structured-errors/error-schema.ts:23]: What would we chose as category for "Wordpress host is down right now and we're getting errors retrieving data" ? Maybe we can have a `THIRD_PARTY` category for that.

Comment (wardpeet) [reply packages/gatsby-cli/src/structured-errors/error-schema.ts:23]: good point, I would have marked that as user

Comment (abhiaiyer91) [reply packages/gatsby-cli/src/structured-errors/error-schema.ts:23]: THIRD_PARTY added

**Comment by abhiaiyer91** [comment]
> Codewise this LGTM overall (Ward left some good comments), but got some questions around this (don't know if it was discussed already):
> 
> 1. We use 5-digit numbers for the errors so that those are better searchable in Google than e.g. `1` as an ID. Can/Should we enforce the same format for plugin errors?
> 2. Since plugins can't override already existing errors from our error map we should at least show a warning somewhere I guess? Otherwise the error reporting silently fails/displays the wrong error message.
> 3. Our error map has a loose format, e.g. all GraphQL errors are [prefixed with 859](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/structured-errors/error-map.ts#L49-L257) and we're currently at 85927.
>    
>    * What happens when a plugin developer decides to claim 85928 and we add the same ID sometime later to our own errorMap?
>    * What happens when two plugins have the same ID?
>    * Should we enforce certain ranges for our internal errorMap and the external?

1. Are people actually searching these? And do we have error explainers like react does? If not and if we're not planning on doing so, whats the point of the ids

2. We can log something

3. 

* Internal Error map wins
* The error maps are prefixed with plugin name so they wont clash like that
* We can but to me the current structure doesnt have rhyme or reason, atleast not ones i see clearly

**Comment by LekoArts** [comment]
1. People are using those IDs for their descriptions in all places and yeah, they also use it for searching for things. We haven't gotten dedicated docs pages for each error but that's something we have as an item on the list for potential next projects

2. 👍 

3.
	- Okay, as long as we have it documented and optionally log something it should be discover-able enough
	- 👍 
	- While the numbers were certainly chosen randomly in the start, the system behind it is something like: For a dedicated area of errors choose the same 4-digits and then count up. So e.g. HTML/SSR errors are [9531x](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/structured-errors/error-map.ts#L21-L34), Webpack errors are [9812x](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/structured-errors/error-map.ts#L35-L48) etc. If we would split up the error 98123 into more errors it would be 98125, 98126 etc.

So overall it's quite arbitrarily chosen and I don't have a strong opinion on the `Should we enforce certain ranges for our internal errorMap and the external?` question, just an idea to minimize the clashing of internal <-> external errorMap :)

File: packages/gatsby-cli/src/structured-errors/error-schema.ts (hunk 11 above)
Comment (wardpeet) [inline packages/gatsby-cli/src/structured-errors/error-schema.ts:23]:
```suggestion
    category: Joi.string().valid([`USER`, `SYSTEM`, `THIRD_PARTY`]),
```

**Review by wardpeet — APPROVED** [review]
LGTM

**Comment by wesrice** [comment]
👋  Hey. This doesn't play well with Gatsby + TypeScript. I'll open a new issue once I understand the problem better.

```
3:31 Cannot find module 'gatsby-cli/src/structured-errors/error-map' or its corresponding type declarations.
    1 | import { Actions, ActivityStatuses, ActivityTypes } from "../constants";
    2 | import { IStructuredError } from "../../structured-errors/types";
  > 3 | import { ErrorCategory } from "gatsby-cli/src/structured-errors/error-map";
      |                               ^
    4 | export interface IGatsbyCLIState {
    5 |     messages: Array<ILog>;
    6 |     activities: {
```

**Comment by abhiaiyer91** [comment]
@wesrice ahhh auto import in vscode. my bad, i have this fixed in this PR https://github.com/gatsbyjs/gatsby/pull/27340

**Pull request merged**